draft-ietf-tls-tls13.md

```diff
@@ -3588,6 +3554,160 @@ appropriate application traffic key as described in {{updating-traffic-keys}}.
+keep state for each outstanding ticket. The server can determine the
+client's view of the age of the ticket by subtracting the ticket's
+"ticket_age_add value" from the "obfuscated_ticket_age" parameter in
+the client's "pre_shared_key" extension. The server can determine the
+"expected arrival time" of the ClientHello as:
+
+~~~~
+    expected_arrival_time = adjusted_creation_time + client's ticket age
+~~~~
+
+For a given storage window, the server implements anti-replay as
+follows.
+
+1. Verify the PSK binder.
+
+2. If the expected_arrival_time is outside the window or the ClientHello
+   matches a known ClientHello then accept the PSK but
+   reject 0-RTT.
+
+3. If the ClientHello matches a known ClientHello then
+   either abort the handshake with an "illegal_parameter" alert
+   or accept the PSK but reject 0-RTT.
+
+4. Otherwise, store the ClientHello as long as its
+   expected_arrival_time is inside the the window and
+   accept 0-RTT.
+
+The server MUST derive the storage key only from validated sections
+of the ClientHello. If the ClientHello contains multiple
+PSK identities, then an attacker can create multiple ClientHellos
+with different binders for the less-preferred identity on the
+assumption that the server will not verify it, thus bypassing
+anti-replay defenses. If the validated binder or the ClientHello.random
+are used for this purpose, then this attack is not possible.
+
+Because this mechanism does not require storing all outstanding
+tickets, it may be easier to implement in distributed systems with
+high rates of resumption and 0-RTT, at the cost of potentially
+weaker anti-replay defense because of the difficulty reliably
+storing and retrieving the received ClientHello messages. In addition,
+servers MAY implement a separate cache for each cluster, thus
+limiting replay to once per cluster. Servers MAY also implement
+data stores with false positives, such as Bloom filters, in
+which case they MUST respond to apparent replay by rejecting
+0-RTT but SHOULD NOT abort the handshake.
+
+Note: When implementations are freshly started, they SHOULD
+reject 0-RTT as long as any portion of their storage window overlaps
+the startup time. Otherwise, they run the risk of accepting
+replays which were originally sent during that period.
+
+
+### Stateless Time-Based Anti-Replay {#stateless-anti-replay}
+
+Finally, the server can implement a very rough anti-replay mechanism
+merely by measuring the mismatch between client and server views of
+time. The server can determine its view of the age of the ticket by
+subtracting the the time the ticket was issued from the current
+time. If the client and server clocks were running at the same rate,
+the client's view of would be shorter than the actual time elapsed on
+the server by a single round trip time.  This difference is comprised
+of the delay in sending the NewSessionTicket message to the client,
+plus the time taken to send the ClientHello to the server.
```

Comment:
It's a little unfortunate to have the discussion of the time calculations both here and in Client Hello Recording, though I don't have an alternative proposal.